src/librustc_typeck/check/method/suggest.rs

```diff
@@ -79,61 +115,31 @@ pub fn report_error<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
+                    } else {
+                        err.span_note(span, &format!("did you mean to write `{0}.{1}`?",
+                                                     expr_string, item_name));
                     }
+                }
+            }
 
-                    // Determine if the field can be used as a function in some way
-                    let field_ty = field.ty(cx, substs);
-
-                    match field_ty.sty {
-                        // Not all of these (e.g. unsafe fns) implement FnOnce
-                        // so we look for these beforehand
-                        ty::TyClosure(..) | ty::TyBareFn(..) => {
-                            span_stored_function!();
-                        }
-                        // If it's not a simple function, look for things which implement FnOnce
-                        _ => {
-                            if let Ok(fn_once_trait_did) =
-                                    cx.lang_items.require(FnOnceTraitLangItem) {
-                                let infcx = fcx.infcx();
-                                infcx.probe(|_| {
-                                    let fn_once_substs =
-                                        Substs::new_trait(vec![infcx.next_ty_var()],
-                                                          Vec::new(),
-                                                          field_ty);
-                                    let trait_ref =
-                                      ty::TraitRef::new(fn_once_trait_did,
-                                                        cx.mk_substs(fn_once_substs));
-                                    let poly_trait_ref = trait_ref.to_poly_trait_ref();
-                                    let obligation = Obligation::misc(span,
-                                                                      fcx.body_id,
-                                                                      poly_trait_ref
-                                                                         .to_predicate());
-                                    let mut selcx = SelectionContext::new(infcx);
-
-                                    if selcx.evaluate_obligation(&obligation) {
-                                        span_stored_function!();
-                                    } else {
-                                        span_did_you_mean!();
-                                    }
-                                });
-                            } else {
-                                span_did_you_mean!();
-                            }
-                        }
+            if is_fn_ty(&rcvr_ty, &fcx, span) {
+                if let Some(expr) = rcvr_expr {
+                    if let Ok (expr_string) = cx.sess.codemap().span_to_snippet(expr.span) {
+                        err.fileline_note(
+                            expr.span,
+                            &format!("{} is a function, perhaps you wish to call it?",
```

Comment:
in case the snippet doesn't exist, we should check if it's an ExprKind::Path, and use the last path segment as the function name. No suggestion in this case.

Also, still needs tests, but then it should be good to go